blk/components/virt.c

```diff
@@ -0,0 +1,345 @@
+    blk_queue_init(&(clients[1].queue_h), (blk_req_queue_t *)blk_req_queue2, (blk_resp_queue_t *)blk_resp_queue2, BLK_QUEUE_SIZE);
+#endif
+
+    // Initialise fixed size memory allocator and datastore
+    datastore_init(&ds, ds_data, sizeof(ds_data_t), ds_nextfree, ds_used, DATASTORE_SIZE);
+    fsmem_init(&fsmem_data, blk_data_driver, BLK_TRANSFER_SIZE, DRV_MAX_DATA_BUFFERS, &fsmem_avail_bitarr, fsmem_avail_bitarr_words, roundup_bits2words64(DRV_MAX_DATA_BUFFERS));
+
+    // Initialise client channels
+    clients[0].ch = CLIENT_CH_1;
+#if BLK_NUM_CLIENTS > 1
+    clients[1].ch = CLIENT_CH_2;
+#endif
+
+    request_mbr();
+}
+
+static void handle_driver() {
+    blk_response_status_t drv_status;
+    uint16_t drv_success_count;
+    uint32_t drv_resp_id;
+
+    while (!blk_resp_queue_empty(&drv_h)) {
+        blk_dequeue_resp(&drv_h, &drv_status, &drv_success_count, &drv_resp_id);
+        
+        ds_data_t cli_data;
+        datastore_retrieve(&ds, drv_resp_id, &cli_data);
+        
+        // Free bookkeeping data structures regardless of success or failure
+        switch(cli_data.code) {
+            case WRITE_BLOCKS:
+                fsmem_free(&fsmem_data, cli_data.drv_addr, cli_data.count);
+                break;
+            case READ_BLOCKS:
+                fsmem_free(&fsmem_data, cli_data.drv_addr, cli_data.count);
+                break;
+            case FLUSH:
+                break;
+            case BARRIER:
+                break;
+        }
+
+        // Get the corresponding client queue handle
+        blk_queue_handle_t h = clients[cli_data.cli_id].queue_h;
+
+        // Drop response if client resp queue is full
+        if (blk_resp_queue_full(&h)) {
+            continue;
+        }
+
+        if (drv_status == SUCCESS) {
+            switch(cli_data.code) {
+                case READ_BLOCKS:
+                    // Invalidate cache
+                    microkit_arm_vspace_data_invalidate(cli_data.drv_addr, cli_data.drv_addr + (BLK_TRANSFER_SIZE * cli_data.count));
+                    // Copy data buffers from driver to client
+                    memcpy((void *)cli_data.cli_addr, (void *)cli_data.drv_addr, BLK_TRANSFER_SIZE * cli_data.count);
+                    blk_enqueue_resp(&h, SUCCESS, drv_success_count, cli_data.cli_req_id);
+                    break;
+                case WRITE_BLOCKS:
+                    blk_enqueue_resp(&h, SUCCESS, drv_success_count, cli_data.cli_req_id);
+                    break;
+                case FLUSH:
+                case BARRIER:
+                    blk_enqueue_resp(&h, SUCCESS, drv_success_count, cli_data.cli_req_id);
+                    break;
+            }
+        } else {
+            // When more error conditions are added, this will need to be updated to a switch statement
+            blk_enqueue_resp(&h, SEEK_ERROR, drv_success_count, cli_data.cli_req_id);
+        }
+        
+        // @ericc: Print out first 512bytes of data for debugging
```

Comment:
If this is useful to keep around, you can have a `debug.h` header that includes this function.